renpy/display/transform.py

```diff
@@ -87,72 +202,336 @@ def limit_angle(n):
+    if x is None:
+        return None
 
-    def __init__(self, name):
-        self.name = name
+    return position(x)
 
-    def __get__(self, instance, owner):
-        return instance.__dict__.get(self.name, None)
 
-    def __set__(self, instance, value):
-        if isinstance(value, str):
-            value = renpy.easy.displayable(value)
+def any_object(x: Any) -> object:
+    return x
 
-        if isinstance(value, Displayable):
-            value = renpy.display.im.unoptimized_texture(value)
 
-            if instance.texture_uniforms is None:
-                instance.texture_uniforms = set()
+def bool_or_none(x: Any) -> bool | None:
+    if x is None:
+        return x
 
-            instance.texture_uniforms.add(self.name)
+    return bool(x)
 
-        instance.__dict__[self.name] = value
 
+def float_or_none(x: Any) -> float | None:
+    if x is None:
+        return x
 
-class TransformState(renpy.object.Object):
-    last_angle = 0.0
-    last_relative_anchorangle = 0.0
-    last_absolute_anchorangle = 0.0
-    last_events = True
+    return float(x)
 
-    available_width = 0
-    available_height = 0
 
-    radius_type = absolute
+class MatrixLike(Protocol):
+    origin: "MatrixLike | None" = None
 
-    radius_sign = 1
-    relative_anchor_radius_sign = 1
-    absolute_anchor_radius_sign = 1
+    def __call__(self, other: "MatrixLike", done: float, /) -> renpy.display.matrix.Matrix: ...
 
-    texture_uniforms = None
 
-    def __init__(self):
-        # Most fields on this object are set by add_property, at the bottom
-        # of this file.
+def matrix_or_none(x: Any) -> MatrixLike | renpy.display.matrix.Matrix | None:
+    if x is None:
+        return None
+    elif callable(x):
+        return x  # type: ignore
+    else:
+        return renpy.display.matrix.Matrix(x)
 
-        # An xpos (etc) inherited from our child overrides an xpos inherited
-        # from an old transform, but not an xpos set in the current transform.
-        #
-        # inherited_xpos stores the inherited_xpos, which is overridden by the
-        # xpos, if not None.
-        self.inherited_xpos = None
-        self.inherited_ypos = None
-        self.inherited_xanchor = None
-        self.inherited_yanchor = None
 
-        # The last angle that was rotated to.
-        self.last_angle = None
+type MeshValue = tuple[int, int] | bool
 
-    def take_state(self, ts):
-        d = self.__dict__
 
-        for k in all_properties:
-            d[k] = getattr(ts, k)
+def mesh_or_none(x: Any) -> MeshValue | None:
+    if x is None:
+        return None
+    elif isinstance(x, tuple):
+        return x
+    else:
+        return bool(x)
+
+
+type MeshPadValue = tuple[int, int] | tuple[int, int, int, int]
+type FitValue = Literal["contain", "cover", "fill", "scale-down", "scale-up"]
+type PerspectiveValue = bool | float | tuple[float, float, float]
+type PointToValue = tuple[float, float, float] | Camera
+
+
+class TransformProperty[T]:
```

Comment:
This is way too magical to be presented without any comments at all. We've gone from straightforward procedural code to something that analyzes types and executes code based on them. That's fine, but it really does need to be called out with documentation and regular comments, rather than added without any comment whatsoever.